File: bbtest/steps/journal_steps.rb
require_relative 'placeholders'

require 'json'

step "transaction of tenant :tenant should be" do |tenant, expectation|
  transaction = JournalHelper.transaction(tenant, @transaction_id)
  expectation = JSON.parse(expectation)

  expect(transaction["id"]).to eq(expectation["id"]) unless expectation["id"].nil?

  expectation["transfers"].each { |e|
    found = false
    transaction["transfers"].each { |t|
      same = true
      same &&= t["id"] == e["id"] unless e["id"].nil?
      same &&= t["credit"] == e["credit"] unless e["credit"].nil?
      same &&= t["debit"] == e["debit"] unless e["debit"].nil?
      same &&= t["valueDate"] == e["valueDate"] unless e["valueDate"].nil?
      same &&= t["amount"] == e["amount"] unless e["amount"].nil?
      same &&= t["currency"] == e["currency"] unless e["currency"].nil?

      if same
        found = true
        break
      end
    }
    raise "#{e} not found in #{transaction}" unless found
  }
end

step "transaction of tenant :tenant should not exist" do |tenant|
  return if @transaction_id.nil?
  data = JournalHelper.transaction(tenant, @transaction_id)
  expect(data).to be_nil, "transaction found #{tenant} #{@transaction_id}"
end

step "transaction of tenant :tenant should exist" do |tenant|
  expect(@transaction_id).not_to be_nil
  data = JournalHelper.transaction(tenant, @transaction_id)
  expect(data).not_to be_nil, "transaction not found #{tenant} #{@transaction_id}"
end

File: bbtest/steps/journal_steps.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from behave import *
import os


@then('transaction {transaction} of tenant {tenant} should be')
def check_transaction_state(context, tenant, transaction):
  filename = os.path.realpath('/data/t_{}/transaction/{}'.format(tenant, transaction))

  assert os.path.isfile(filename) is True, 'file {} does not exists'.format(filename)

  actual = dict()
  with open(filename, 'r') as fd:
    lines = fd.readlines()
    lines = [line.strip() for line in lines]
    state = lines[0]
    transfers = []

    for transfer in lines[1:]:
      item = transfer.split(' ')

      transfers.append({
        'id': item[0],
        'credit': {
          'name': item[2],
          'tenant': item[1],
        },
        'debit':{
          'name': item[4],
          'tenant': item[3],
        },
        'amount': item[6],
        'currency': item[7],
        'valueDate': item[5],
      })

    actual.update({
      'state': state,
      'transfers': transfers,
    })

  for row in context.table:
    assert row['key'] in actual, 'key {} not found in {}'.format(row['key'], actual)
    assert actual[row['key']] == row['value'], 'extected {} got {} instead'.format(row['key'], actual)


@then('transaction of tenant {tenant} should not exist')
def transaction_should_not_exist(context, tenant):
  if context.last_transaction_id:
    path = '/data/t_{}/transaction/{}'.format(tenant, context.last_transaction_id)
    assert os.path.isfile(path) is False, "{} exists but should not".format(path)


@then('transaction of tenant {tenant} should exist')
def transaction_should_exist(context, tenant):
  assert context.last_transaction_id, 'missing last_transaction_id from context'
  path = '/data/t_{}/transaction/{}'.format(tenant, context.last_transaction_id)
  assert os.path.isfile(path) is True, "{} does not exists but should".format(path)
